### PR TITLE
Fix compilation warnings with SUNDIALS 7.1.0

### PR DESCRIPTION
**Pull request by Steven-Roberts**
In SUNDIALS 7.1.0, many of the functions that start with `ARKstep` were deprecated and renamed with `ARKode` prefix. This fixes the compilations warnings.

Tested with SUNDIALS versions 5.8.0, 6.7.0, 7.0.0, and 7.1.0.

### CODE REVIEW

File: src/solver/impls/arkode/arkode.cxx [hunk 1]
```diff
@@ -213,51 +215,28 @@ int ArkodeSolver::init() {
...
-  case Treatment::Implicit:
-    output_info.write("\tUsing ARKStep Implicit solver \n");
-    if (ARKStepSetImplicit(arkode_mem) != ARK_SUCCESS) {
-      throw BoutException("ARKStepSetImplicit failed\n");
-    }
-    break;
-  default:
-    throw BoutException("Invalid treatment: {}\n", toString(treatment));
-  }
-
```
Comment (Steven-Roberts) [inline src/solver/impls/arkode/arkode.cxx:238]: @drreynolds @maggul This looks redundant with the call to `ARKstepCreate` before this, so I've removed the code. Can you double check this is ok?

Comment (drreynolds) [reply src/solver/impls/arkode/arkode.cxx:238]: I agree with @Steven-Roberts -- there is no need to call `SetImEx`, `SetExplicit`, or `SetImplicit` in this context.

File: include/bout/sundials_backports.hxx [hunk 2]
```diff
@@ -21,11 +21,23 @@
...
+#error "Unable to determine SUNDIALS version"
+#endif
+
+#define SUNDIALS_VERSION_AT_LEAST(major, minor, patch) \
+  (major < SUNDIALS_VERSION_MAJOR                      \
+   || (major == SUNDIALS_VERSION_MAJOR                 \
+       && (minor < SUNDIALS_VERSION_MINOR              \
+           || (minor == SUNDIALS_VERSION_MINOR && patch <= SUNDIALS_VERSION_PATCH))))
+#define SUNDIALS_VERSION_LESS_THAN(major, minor, patch) \
+  (!SUNDIALS_VERSION_AT_LEAST(major, minor, patch))
```
Comment (ZedThree) [inline include/bout/sundials_backports.hxx:34]: To keep clang-tidy happy, please could you make the following changes:

```suggestion
// NOLINTBEGIN(cppcoreguidelines-macro-usage)
#define SUNDIALS_VERSION_AT_LEAST(major, minor, patch) \
  ((major) < SUNDIALS_VERSION_MAJOR                      \
   || ((major) == SUNDIALS_VERSION_MAJOR                 \
       && ((minor) < SUNDIALS_VERSION_MINOR              \
           || ((minor) == SUNDIALS_VERSION_MINOR && (patch) <= SUNDIALS_VERSION_PATCH))))
#define SUNDIALS_VERSION_LESS_THAN(major, minor, patch) \
  (!SUNDIALS_VERSION_AT_LEAST(major, minor, patch))
// NOLINTEND(cppcoreguidelines-macro-usage)
```

File: src/solver/impls/arkode/arkode.cxx [hunk 3]
```diff
@@ -213,51 +215,28 @@ int ArkodeSolver::init() {
...
   // For callbacks, need pointer to solver object
-  if (ARKStepSetUserData(arkode_mem, this) != ARK_SUCCESS) {
-    throw BoutException("ARKStepSetUserData failed\n");
+  if (ARKodeSetUserData(arkode_mem, this) != ARK_SUCCESS) {
+    throw BoutException("ARKodeSetUserData failed\n");
   }
 
-  if (ARKStepSetLinear(arkode_mem, set_linear) != ARK_SUCCESS) {
-    throw BoutException("ARKStepSetLinear failed\n");
+  if (ARKodeSetLinear(arkode_mem, set_linear) != ARK_SUCCESS) {
```
Comment (ZedThree) [inline src/solver/impls/arkode/arkode.cxx:223]: Another bit just to keep clang-tidy happy:

```suggestion
  if (ARKodeSetLinear(arkode_mem, static_cast<int>(set_linear)) != ARK_SUCCESS) {
```

**Review by ZedThree — CHANGES_REQUESTED** [review]
LGTM, thanks @Steven-Roberts! Please could you just fix a couple of suggestions from clang-tidy?

**Comment by Steven-Roberts** [comment]
@ZedThree @bendudson I think this should be ready to merge. Let me know if anything else needs to be done.

**Comment by Steven-Roberts** [comment]
Whoops, that's right. I'll see if it passes now.

**Review by dschwoerer — APPROVED** [review]
CI failures are due to permission errors, as it runs from a fork.

Everything else looks good :+1:

**Pull request merged**